### ISSUE TITLE
Line layout, protrusion

### ISSUE THREAD

**Issue by bdarcus** [typst/typst#261]
I'm new to typst, so forgive me if I'm missing something.

Could you add more advanced, microtype/InDesign-like, things like protrusion?

Using Minion Pro, `typst`:

![typst-prot](https://user-images.githubusercontent.com/1134/227565633-8fd8008b-bb4c-469b-8eb0-b933a0faac01.png)

Here's lualatex with microtype:

![tex-protrusion](https://user-images.githubusercontent.com/1134/227565693-55da4313-5c51-4a3c-b0a9-365cdd8fc1d3.png)

**Comment by hirschenberger** [comment]
I'd like to implement the protrusion functionality. I did a bit of research and it seems even in texmacs and pdftex it's implemented with a lookup table `Symbol -> Shift`. Regardless of the font used.

I wonder why it's done in such a simple and uncomplete way. Wouldn't it be possible to directly calculate glyph metrics like `blackness`, `center of mass` .... to align the border more precisely?

**Comment by bdarcus** [comment]
> I wonder why it's done in such a simple and uncomplete way.

I don't know.  Maybe it wasn't convenient or possible to do that when it was first implemented?

Have you checked the [luatex repo](https://github.com/search?q=repo%3ATeX-Live%2Fluatex%20protrusion%20&type=code)? Perhaps they asked the same question when working on it?

**Comment by Enivex** [comment]
> I'd like to implement the protrusion functionality. I did a bit of research and it seems even in texmacs and pdftex it's implemented with a heuristic lookup table `Symbol -> Shift`. Regardless of the font used.
> 
> A good source is the [dissertation](https://www.tug.org/TUGboat/tb21-4/tb69thanh.pdf) of Hàn Thế Thành, who wrote pdftex.
> 
> I wonder why it's done in such a simple and uncomplete way. Wouldn't it be possible to directly calculate glyph metrics like e.g. `blackness`, `center of mass` to align the border more precisely?

I would look to the microtype package, which is the one that implements the advanced features

https://ctan.org/pkg/microtype?lang=en

There's a 270 page document on implementation details.

**Comment by hirschenberger** [comment]
> I would look to the microtype package, which is the one that implements the advanced features
> 
> https://ctan.org/pkg/microtype?lang=en
> 
> There's a 270 page document on implementation details.

Yes, I know that, but on page 148 the author mentions that his protusion is based on the tables that Hàn Thế Thành created in the dissertiation I mentioned above.

**Comment by bdarcus** [comment]
In any case, I don't claim to understand the nuances here, but I agree with you @hirschenberger that it'd be ideal to calculate the metrics rather than assume some fixed table.

**Comment by bdarcus** [comment]
Was curious, and found this:

https://stackoverflow.com/questions/68151488/rusttype-get-text-width-for-font

**Comment by Enivex** [comment]
> The typesetting in the original image looks quite good. For me it looks much worse:
> 
> ![test](https://user-images.githubusercontent.com/107212151/228559648-e3ee0364-c0d7-44b3-95d8-30e641edee18.PNG)
> 
> The problem in the algorithm used here is that when some line is too long or too short (which is not really that problematic),
> the line is just broken and everything is kept like it is. That's a very fast algorithm.
> But what is better is this [one](https://en.wikipedia.org/wiki/Microtypography#/media/File:Three_images_of_a_paragraph_from_The_Red_Badge_of_Courage_(Stephen_Crane,_1895)_to_illustrate_kerning,_ligatures,_hyphenation_and_microtypography.png) (top image). 
> Instead of just breaking the line, add to all horizontal spacing within the line the same amount of additional horizontal space,
> so that the space at the end is now zero.
> With this new algorithm we will achieve the top picture in the linked wikipedia image instead of the one I linked.
> I think, this is the most important first step.

That doesn't look right. I think justification is disabled, either by accident or some bug.

**Comment by laurmaedje** [comment]
Typst already does this, but it doesn't let all characters overhang by the same amount. You can see the ratios [here](https://github.com/typst/typst/blob/10a3fbd174fc1a3f95937c9120d478c8e94c277d/library/src/layout/par.rs#L1362). The behaviour can be disabled with [this property](http://localhost:8081/docs/reference/text/text/#parameters--overhang). Could be made more configurable so that you can also have full overhang.

**Comment by bdarcus** [comment]
> Typst already does this, but it doesn't let all characters overhang by the same amount.

Oh good, so the functionality is there; just a question of optimizing, and/or making it configurable, and/or automating optimization based on [calculating the glyph metrics](https://github.com/typst/typst/issues/261#issuecomment-1483756952)?

In the two examples I posted, the tex/microtype one clearly looks better.

**Comment by hirschenberger** [comment]
I'm currently working on a PR with a more complete protrusion table based on the values texmacs uses.

**Comment by bdarcus** [comment]
Am I wrong, or did this get fixed at some point, so that you can close this?

**Comment by Enivex** [comment]
> Am I wrong, or did this get fixed at some point, so that you can close this?

It was never broken. I think it's still open because it can still be improved.

**Comment by hurzl** [comment]
With LaTeX's microtype package 
1. you can define protrusion amounts individually per symbol, and
2. protrusion is also applied to the left side.

**Comment by adamkrellenstein** [comment]
In case it's useful: as of 0.15.0 `text.overhang` is still a bool, and the overhang amounts are a hardcoded table in `typst-layout/src/inline/line.rs` covering only dashes, hyphens, `.,:;` and two Arabic marks — letters get 0, so caps like W/A/V/T at a flush-left line start don't protrude. Looks like #7782 would make this configurable per-glyph (incl. by glyph name) on both margins.
